File: windows/WeChatExporter.Windows/Services/ChatJsonProcessor.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace WeChatExporter.Services;

public sealed record ExportPreviewResult(
    int ContactCount,
    int MessageCount,
    int MediaCount,
    long EstimatedBytes,
    IReadOnlyDictionary<string, int> ByType)
{
    public string EstimatedSizeText
    {
        get
        {
            if (EstimatedBytes < 1024) return $"{EstimatedBytes} B";
            if (EstimatedBytes < 1024 * 1024) return $"{EstimatedBytes / 1024.0:F1} KB";
            if (EstimatedBytes < 1024L * 1024L * 1024L) return $"{EstimatedBytes / 1024.0 / 1024.0:F1} MB";
            return $"{EstimatedBytes / 1024.0 / 1024.0 / 1024.0:F2} GB";
        }
    }

    public string SummaryText
    {
        get
        {
            var typeLine = string.Join(" · ", ByType.OrderBy(kv => kv.Key).Select(kv => $"{kv.Key} {kv.Value}"));
            var baseLine = $"会话 {ContactCount} 个 · 消息约 {MessageCount} 条 · 媒体约 {MediaCount} 个 · 预估 {EstimatedSizeText}";
            return string.IsNullOrWhiteSpace(typeLine) ? baseLine : $"{baseLine}\n类型分布：{typeLine}";
        }
    }
}

public sealed class ChatJsonFilterOptions
{
    public long? SinceUnix { get; init; }
    public long? UntilUnix { get; init; }
    public HashSet<string> EnabledTypes { get; init; } =
        MessageTypeFilterInfo.AllIds().ToHashSet(StringComparer.OrdinalIgnoreCase);

    public bool FilterTypes
    {
        get
        {
            if (EnabledTypes.Count == 0) return false;
            var all = MessageTypeFilterInfo.AllIds();
            return EnabledTypes.Count < all.Count || all.Any(id => !EnabledTypes.Contains(id));
        }
    }
}

/// <summary>Post-processes wx-cli chat.json exports without changing the talker metadata.</summary>
public static class ChatJsonProcessor
{
    private static readonly JsonSerializerOptions WriteOptions = new()
    {
        WriteIndented = true
    };

    public static int ApplyFilters(string jsonPath, ChatJsonFilterOptions options, Action<string>? log = null)
    {
        if (!System.IO.File.Exists(jsonPath)) return 0;
        var root = JsonNode.Parse(System.IO.File.ReadAllText(jsonPath));
        if (root is null) return 0;

        var array = ExtractRows(root, out _);
        if (array is null) return 0;

        var original = array.Count;
        var kept = new List<JsonNode?>();
        foreach (var row in array)
        {
            if (row is not null && Passes(row, options))
                kept.Add(row.DeepClone());
        }

        if (kept.Count != original)
        {
            array.Clear();
            foreach (var row in kept)
                array.Add(row);
            System.IO.File.WriteAllText(jsonPath, root.ToJsonString(WriteOptions), Encoding.UTF8);
            log?.Invoke($"已按设置过滤消息：{original} → {kept.Count}");
        }

        RewriteTxt(System.IO.Path.Combine(System.IO.Path.GetDirectoryName(jsonPath) ?? "", "chat.txt"), kept);
        return kept.Count;
    }

    public static void ApplyNicknameMap(string jsonPath, IReadOnlyDictionary<string, string> map, Action<string>? log = null)
    {
        if (map.Count == 0 || !System.IO.File.Exists(jsonPath)) return;
        var root = JsonNode.Parse(System.IO.File.ReadAllText(jsonPath));
        if (root is null) return;
        var array = ExtractRows(root, out _);
        if (array is null) return;

        var changed = 0;
        foreach (var row in array)
        {
            if (row is not JsonObject obj) continue;
            var nested = obj["message"] as JsonObject;
            var raw = GetString(obj, "sender", "from") ?? (nested is null ? null : GetString(nested, "sender", "from"));
            if (string.IsNullOrWhiteSpace(raw)) continue;

            if (!map.TryGetValue(raw, out var name)
                && !map.TryGetValue(raw.ToLowerInvariant(), out name))
                continue;
            if (string.IsNullOrWhiteSpace(name)) continue;

            if (string.IsNullOrWhiteSpace(GetString(obj, "sender_display_name")))
            {
                obj["sender_display_name"] = name;
                changed++;
            }
            if (nested is not null && string.IsNullOrWhiteSpace(GetString(nested, "sender_display_name")))
                nested["sender_display_name"] = name;
        }

        if (changed > 0)
        {
            System.IO.File.WriteAllText(jsonPath, root.ToJsonString(WriteOptions), Encoding.UTF8);
            log?.Invoke($"已应用群成员昵称映射 {changed} 处");
        }
    }

    public static ExportPreviewResult Preview(string jsonPath, string? sourceDir = null)
    {
        if (!System.IO.File.Exists(jsonPath))
            return new ExportPreviewResult(1, 0, 0, 0, new Dictionary<string, int>());

        var dataBytes = new System.IO.FileInfo(jsonPath).Length;
        var root = JsonNode.Parse(System.IO.File.ReadAllText(jsonPath));
        if (root is null)
            return new ExportPreviewResult(1, 0, 0, dataBytes, new Dictionary<string, int>());

        var array = ExtractRows(root, out _);
        if (array is null)
            return new ExportPreviewResult(1, 0, 0, dataBytes, new Dictionary<string, int>());

        var byType = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        var mediaCount = 0;
        var bytes = dataBytes;
        var baseDir = sourceDir ?? System.IO.Path.GetDirectoryName(jsonPath) ?? "";

        foreach (var row in array)
        {
            if (row is null) continue;
            var source = (row["message"] as JsonObject) ?? row;
            var msgType = GetInt(source, "msg_type", "type") ?? GetInt(row, "msg_type", "type");
            var typeName = GetString(row, "type_name", "type") ?? GetString(source, "type_name", "type");
            var label = MessageTypeFilterInfo.Matching(msgType, typeName) is { } filter
                ? MessageTypeFilterInfo.Title(filter)
                : "其他";
            byType[label] = byType.TryGetValue(label, out var n) ? n + 1 : 1;

            foreach (var media in MediaFiles(row))
            {
                mediaCount++;
                var path = ResolveMediaPath(media, baseDir);
                bytes += System.IO.File.Exists(path) ? new System.IO.FileInfo(path).Length : 80_000;
            }
        }

        return new ExportPreviewResult(1, array.Count, mediaCount, bytes, byType);
    }

    public static long? LatestCreateTime(string jsonPath)
    {
        if (!System.IO.File.Exists(jsonPath)) return null;
        var root = JsonNode.Parse(System.IO.File.ReadAllText(jsonPath));
        if (root is null) return null;
        var array = ExtractRows(root, out _);
        if (array is null) return null;

        long max = 0;
        foreach (var row in array)
        {
            if (row is null) continue;
            var source = (row["message"] as JsonObject) ?? row;
            var ts = GetLong(source, "create_time", "timestamp") ?? GetLong(row, "create_time", "timestamp");
            if (ts is long value)
                max = Math.Max(max, NormalizeUnix(value));
        }
        return max > 0 ? max : null;
    }

    public static void InjectVoiceTranscripts(string jsonPath, IReadOnlyDictionary<string, string> transcripts, Action<string>? log = null)
    {
        if (transcripts.Count == 0 || !System.IO.File.Exists(jsonPath)) return;
        var root = JsonNode.Parse(System.IO.File.ReadAllText(jsonPath));
        if (root is null) return;
        var array = ExtractRows(root, out _);
        if (array is null) return;

        var hit = 0;
        foreach (var row in array)
        {
            if (row is not JsonObject obj) continue;
            var nested = obj["message"] as JsonObject;
            foreach (var media in MediaFiles(obj))
            {
                var name = System.IO.Path.GetFileName(media);
                if (!transcripts.TryGetValue(media, out var text)
                    && !transcripts.TryGetValue(name, out text))
                    continue;

                var existing = GetString(obj, "content") ?? (nested is null ? null : GetString(nested, "content")) ?? "";
                var prefix = "[语音转写] ";
                var merged = existing.Contains(prefix, StringComparison.Ordinal)
                    ? existing
                    : string.IsNullOrEmpty(existing) ? prefix + text : existing + "\n" + prefix + text;
                obj["content"] = merged;
                obj["snippet"] = merged;
                if (nested is not null)
                    nested["content"] = merged;
                hit++;
                break;
            }
        }

        if (hit > 0)
        {
            System.IO.File.WriteAllText(jsonPath, root.ToJsonString(WriteOptions), Encoding.UTF8);
            log?.Invoke($"已写入 {hit} 条语音转写");
        }
    }

    private static bool Passes(JsonNode row, ChatJsonFilterOptions options)
    {
        var source = (row["message"] as JsonObject) ?? row;
        var ts = GetLong(source, "create_time", "timestamp") ?? GetLong(row, "create_time", "timestamp");
        if (ts is long rawTs)
        {
            var normalized = NormalizeUnix(rawTs);
            if (options.SinceUnix is long since && normalized < since) return false;
            if (options.UntilUnix is long until && normalized > until) return false;
        }

        if (options.FilterTypes)
        {
            var msgType = GetInt(source, "msg_type", "type") ?? GetInt(row, "msg_type", "type");
            var typeName = GetString(row, "type_name", "type") ?? GetString(source, "type_name", "type");
            var matched = MessageTypeFilterInfo.Matching(msgType, typeName);
            if (matched is null)
                return options.EnabledTypes.Contains(MessageTypeFilterInfo.Id(MessageTypeFilter.App));
            return options.EnabledTypes.Contains(MessageTypeFilterInfo.Id(matched.Value));
        }
        return true;
    }

    private static JsonArray? ExtractRows(JsonNode root, out string? container)
    {
        container = null;
        if (root is JsonArray rootArray)
            return rootArray;

        if (root is JsonObject obj)
        {
            foreach (var key in new[] { "items", "messages", "results" })
            {
                if (obj[key] is JsonArray array)
                {
                    container = key;
                    return array;
                }
            }
        }
        return null;
    }

    private static IEnumerable<string> MediaFiles(JsonNode row)
    {
        var source = (row["message"] as JsonObject) ?? row;
        if (row["media_files"] is JsonArray media)
        {
            foreach (var item in media)
            {
                var value = item?.GetValue<string>();
                if (!string.IsNullOrWhiteSpace(value)) yield return value;
            }
        }
        else if (source["media_files"] is JsonArray nestedMedia)
        {
            foreach (var item in nestedMedia)
            {
                var value = item?.GetValue<string>();
                if (!string.IsNullOrWhiteSpace(value)) yield return value;
            }
        }
    }

    private static void RewriteTxt(string txtPath, IReadOnlyList<JsonNode?> rows)
    {
        if (string.IsNullOrWhiteSpace(txtPath)) return;
        var lines = new List<string>();
        foreach (var row in rows)
        {
            if (row is null) continue;
            var source = (row["message"] as JsonObject) ?? row;
            var time = GetString(row, "time", "timestamp_str")
                       ?? GetString(source, "time", "timestamp_str")
                       ?? "";
            var sender = GetString(row, "sender_display_name", "sender", "from")
                         ?? GetString(source, "sender_display_name", "sender")
                         ?? "";
            var content = GetString(row, "snippet", "content", "text")
                          ?? GetString(source, "snippet", "content", "text")
                          ?? "";
            lines.Add($"[{time}] {sender}: {content}");
        }
        System.IO.File.WriteAllText(txtPath, string.Join(Environment.NewLine, lines), Encoding.UTF8);
    }

    private static string ResolveMediaPath(string media, string baseDir)
    {
        if (System.IO.Path.IsPathRooted(media)) return media;
        return System.IO.Path.Combine(baseDir, media.Replace('/', System.IO.Path.DirectorySeparatorChar));
    }

    private static string? GetString(JsonNode? node, params string[] keys)
    {
        if (node is not JsonObject obj) return null;
        foreach (var key in keys)
        {
            var value = obj[key];
            if (value is null) continue;
            if (value.GetValueKind() == JsonValueKind.String)
            {
                var s = value.GetValue<string>();
                if (!string.IsNullOrEmpty(s)) return s;
            }
            else if (value.GetValueKind() == JsonValueKind.Number)
            {
                return value.ToJsonString();
            }
        }
        return null;
    }

    private static int? GetInt(JsonNode? node, params string[] keys)
    {
        var value = GetLong(node, keys);
        if (value is null) return null;
        return value > int.MaxValue || value < int.MinValue ? null : (int)value.Value;
    }

    private static long? GetLong(JsonNode? node, params string[] keys)
    {
        if (node is not JsonObject obj) return null;
        foreach (var key in keys)
        {
            var value = obj[key];
            if (value is null) continue;
            if (value.GetValueKind() == JsonValueKind.Number)
            {
                try { return value.GetValue<long>(); }
                catch { return null; }
            }
            if (value.GetValueKind() == JsonValueKind.String && long.TryParse(value.GetValue<string>(), out var parsed))
                return parsed;
        }
        return null;
    }

    private static long NormalizeUnix(long ts) => ts > 9_999_999_999 ? ts / 1000 : ts;
}

File: Sources/WeChatExporter/Services/ChatJsonProcessor.swift
import Foundation

/// 对 CLI 导出的 chat.json 做时间/类型过滤、昵称映射、预览统计。
enum ChatJsonProcessor {
    struct FilterOptions {
        var sinceUnix: Int?
        var untilUnix: Int?
        var enabledTypes: Set<MessageTypeFilter>
        /// 空集合 = 不过滤类型。
        var filterTypes: Bool { !enabledTypes.isEmpty && enabledTypes.count < MessageTypeFilter.allCases.count }
    }

    @discardableResult
    static func applyFilters(to jsonURL: URL, options: FilterOptions, log: ((String) -> Void)? = nil) throws -> Int {
        guard FileManager.default.fileExists(atPath: jsonURL.path) else { return 0 }
        let data = try Data(contentsOf: jsonURL)
        let root = try JSONSerialization.jsonObject(with: data)
        let (container, rows) = try extractRows(from: root)
        let filtered = rows.filter { row in
            passes(row: row, options: options)
        }
        if filtered.count == rows.count { return filtered.count }

        let rewritten = rewrite(root: root, container: container, rows: filtered)
        let out = try JSONSerialization.data(withJSONObject: rewritten, options: [.prettyPrinted, .sortedKeys])
        try out.write(to: jsonURL, options: .atomic)
        log?("已按设置过滤消息：\(rows.count) → \(filtered.count)")
        rewriteTxtIfNeeded(alongside: jsonURL, rows: filtered)
        return filtered.count
    }

    static func applyNicknameMap(to jsonURL: URL, map: [String: String], log: ((String) -> Void)? = nil) throws {
        guard !map.isEmpty, FileManager.default.fileExists(atPath: jsonURL.path) else { return }
        let data = try Data(contentsOf: jsonURL)
        let root = try JSONSerialization.jsonObject(with: data)
        let (container, rows) = try extractRows(from: root)
        var changed = 0
        let mapped = rows.map { row -> [String: Any] in
            var copy = row
            var nested = row["message"] as? [String: Any]
            let senderKeys = ["sender", "from"]
            for key in senderKeys {
                if let raw = (copy[key] as? String) ?? (nested?[key] as? String),
                   let nice = map[raw] ?? map[raw.lowercased()],
                   !nice.isEmpty {
                    if copy["sender_display_name"] == nil || (copy["sender_display_name"] as? String)?.isEmpty == true {
                        copy["sender_display_name"] = nice
                        changed += 1
                    }
                    if var n = nested {
                        if n["sender_display_name"] == nil {
                            n["sender_display_name"] = nice
                            nested = n
                        }
                    }
                }
            }
            if let nested { copy["message"] = nested }
            return copy
        }
        let rewritten = rewrite(root: root, container: container, rows: mapped)
        let out = try JSONSerialization.data(withJSONObject: rewritten, options: [.prettyPrinted, .sortedKeys])
        try out.write(to: jsonURL, options: .atomic)
        if changed > 0 { log?("已应用群成员昵称映射 \(changed) 处") }
    }

    static func preview(from jsonURL: URL, sourceDir: URL? = nil) -> ExportPreviewResult {
        guard let data = try? Data(contentsOf: jsonURL),
              let root = try? JSONSerialization.jsonObject(with: data),
              let (_, rows) = try? extractRows(from: root) else {
            return ExportPreviewResult(contactCount: 1, messageCount: 0, mediaCount: 0, estimatedBytes: 0, byType: [:])
        }

        var byType: [String: Int] = [:]
        var mediaCount = 0
        var bytes: Int64 = Int64(data.count)
        for row in rows {
            let source = (row["message"] as? [String: Any]) ?? row
            let msgType = intField(source, keys: ["msg_type", "type"]) ?? intField(row, keys: ["msg_type", "type"])
            let typeName = stringField(row, keys: ["type_name"]) ?? stringField(source, keys: ["type_name"])
            let label = MessageTypeFilter.matching(msgType: msgType, typeName: typeName)?.title ?? "其他"
            byType[label, default: 0] += 1
            let media = (row["media_files"] as? [String]) ?? (source["media_files"] as? [String]) ?? []
            mediaCount += media.count
            for path in media {
                let url = resolveMedia(path, sourceDir: sourceDir ?? jsonURL.deletingLastPathComponent())
                if let attrs = try? FileManager.default.attributesOfItem(atPath: url.path),
                   let size = attrs[.size] as? NSNumber {
                    bytes += size.int64Value
                } else {
                    bytes += 80_000
                }
            }
        }
        return ExportPreviewResult(
            contactCount: 1,
            messageCount: rows.count,
            mediaCount: mediaCount,
            estimatedBytes: bytes,
            byType: byType
        )
    }

    static func latestCreateTime(in jsonURL: URL) -> Int? {
        guard let data = try? Data(contentsOf: jsonURL),
              let root = try? JSONSerialization.jsonObject(with: data),
              let (_, rows) = try? extractRows(from: root) else { return nil }
        var maxTs = 0
        for row in rows {
            let source = (row["message"] as? [String: Any]) ?? row
            if let ts = intField(source, keys: ["create_time", "timestamp"]) ?? intField(row, keys: ["create_time", "timestamp"]) {
                maxTs = max(maxTs, ts)
            }
        }
        return maxTs > 0 ? maxTs : nil
    }

    static func injectVoiceTranscripts(in jsonURL: URL, transcripts: [String: String], log: ((String) -> Void)? = nil) throws {
        guard !transcripts.isEmpty else { return }
        let data = try Data(contentsOf: jsonURL)
        let root = try JSONSerialization.jsonObject(with: data)
        let (container, rows) = try extractRows(from: root)
        var hit = 0
        let mapped = rows.map { row -> [String: Any] in
            var copy = row
            var nested = row["message"] as? [String: Any]
            let source = nested ?? row
            let media = (row["media_files"] as? [String]) ?? (source["media_files"] as? [String]) ?? []
            for path in media {
                let name = URL(fileURLWithPath: path).lastPathComponent
                if let text = transcripts[name] ?? transcripts[path] {
                    let prefix = "[语音转写] "
                    let existing = (copy["content"] as? String) ?? (nested?["content"] as? String) ?? ""
                    let merged = existing.contains(prefix) ? existing : (existing.isEmpty ? prefix + text : existing + "\n" + prefix + text)
                    copy["content"] = merged
                    copy["snippet"] = merged
                    if var n = nested {
                        n["content"] = merged
                        nested = n
                    }
                    hit += 1
                    break
                }
            }
            if let nested { copy["message"] = nested }
            return copy
        }
        let rewritten = rewrite(root: root, container: container, rows: mapped)
        let out = try JSONSerialization.data(withJSONObject: rewritten, options: [.prettyPrinted, .sortedKeys])
        try out.write(to: jsonURL, options: .atomic)
        if hit > 0 { log?("已写入 \(hit) 条语音转写") }
    }

    // MARK: - internals

    private static func passes(row: [String: Any], options: FilterOptions) -> Bool {
        let source = (row["message"] as? [String: Any]) ?? row
        if let since = options.sinceUnix, let ts = intField(source, keys: ["create_time", "timestamp"]) ?? intField(row, keys: ["create_time", "timestamp"]), ts < since {
            return false
        }
        if let until = options.untilUnix, let ts = intField(source, keys: ["create_time", "timestamp"]) ?? intField(row, keys: ["create_time", "timestamp"]), ts > until {
            return false
        }
        if options.filterTypes {
            let msgType = intField(source, keys: ["msg_type", "type"]) ?? intField(row, keys: ["msg_type", "type"])
            let typeName = stringField(row, keys: ["type_name"]) ?? stringField(source, keys: ["type_name"])
            guard let matched = MessageTypeFilter.matching(msgType: msgType, typeName: typeName) else {
                return options.enabledTypes.contains(.app) // 未知类型归入 app/其他时保留若选了链接
            }
            return options.enabledTypes.contains(matched)
        }
        return true
    }

    private static func extractRows(from root: Any) throws -> (container: String?, rows: [[String: Any]]) {
        if let array = root as? [[String: Any]] {
            return (nil, array)
        }
        if let dict = root as? [String: Any] {
            for key in ["items", "messages", "results"] {
                if let items = dict[key] as? [[String: Any]] {
                    return (key, items)
                }
            }
        }
        throw AppError.exportFailed("chat.json 中未找到消息列表")
    }

    private static func rewrite(root: Any, container: String?, rows: [[String: Any]]) -> Any {
        if container == nil { return rows }
        guard var dict = root as? [String: Any], let key = container else { return rows }
        dict[key] = rows
        return dict
    }

    private static func rewriteTxtIfNeeded(alongside jsonURL: URL, rows: [[String: Any]]) {
        let txt = jsonURL.deletingLastPathComponent().appendingPathComponent("chat.txt")
        var lines: [String] = []
        for row in rows {
            let source = (row["message"] as? [String: Any]) ?? row
            let time = stringField(row, keys: ["time", "timestamp_str"])
                ?? stringField(source, keys: ["time", "timestamp_str"])
                ?? ""
            let sender = stringField(row, keys: ["sender_display_name", "sender", "from"])
                ?? stringField(source, keys: ["sender_display_name", "sender"])
                ?? ""
            let content = stringField(row, keys: ["snippet", "content", "text"])
                ?? stringField(source, keys: ["snippet", "content", "text"])
                ?? ""
            lines.append("[\(time)] \(sender): \(content)")
        }
        try? lines.joined(separator: "\n").write(to: txt, atomically: true, encoding: .utf8)
    }

    private static func resolveMedia(_ path: String, sourceDir: URL) -> URL {
        if path.hasPrefix("/") { return URL(fileURLWithPath: path) }
        return sourceDir.appendingPathComponent(path)
    }

    private static func stringField(_ row: [String: Any], keys: [String]) -> String? {
        for key in keys {
            if let value = row[key] as? String, !value.isEmpty { return value }
        }
        return nil
    }

    private static func intField(_ row: [String: Any], keys: [String]) -> Int? {
        for key in keys {
            if let value = row[key] as? Int { return value }
            if let value = row[key] as? Int64 { return Int(value) }
            if let value = row[key] as? Double { return Int(value) }
        }
        return nil
    }
}
